### PR TITLE
CLN refactor cell_set

### PR DESCRIPTION
**Pull request by pierreglaser**
This PR:

- documents clearly the hack used to set the contents of a cell before [PEP 3104](https://www.python.org/dev/peps/pep-3104/)
- does not use this hack but simply set the `cell_contents` of the cell after  https://bugs.python.org/issue30486 

By the way this is probably the craziest python code I have ever seen in my entire life.

### CODE REVIEW

File: cloudpickle/cloudpickle.py [hunk 1]
```diff
@@ -269,62 +269,75 @@ def func():
...
+
+    In Python3.7, cell_contents is writeable, so setting the contents of a cell
+    can be done simply using
+    >>> cell.cell_contents = value
+
+    In earlier Python3 versions, the cell_contents attribute of a cell is read
+    only, but this limitation can be worked around by leveraging the Python 3
+    ``nonlocal`` keyword.
+
+    In Python2 however, this attribute is read only, and there is no non-local
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]:
```suggestion
    In Python2 however, this attribute is read only, and there is no ``nonlocal``
```

File: cloudpickle/cloudpickle.py [hunk 2]
```diff
@@ -269,62 +269,75 @@ def func():
...
+    load a local variable var and increment it. As g does not assign the local
+    variable ``var`` any initial value, calling f(1)() will fail at runtime.)
+
+    Our objective is to set the value of a given cell ``cell``. So we need to
+    somewhat reference our ``cell`` object into the ``inner`` function so that
+    this object (and not the smoke cell of the lambda function) gets affected
+    by the STORE_DEREF operation.
+
+    In inner, ``cell`` is referenced as a cell variable (an enclosing variable
+    that is referenced by inner function). If we create a new function cell_set
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]: by the inner function

File: cloudpickle/cloudpickle.py [hunk 3]
```diff
@@ -354,23 +367,14 @@ def inner(value):
                 co.co_name,
                 co.co_firstlineno,
                 co.co_lnotab,
-                co.co_cellvars,  # this is the trickery
-                (),
+                co.co_cellvars,  # co_freevars is initialized with co_cellvars
+                (),  # co_cellvars is made empty
             )
-
-_cell_set_template_code = _make_cell_set_template_code()
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]: Why not keep this? Creating a new `_cell_set_template_code` instance each time we call `_cell_set` might lead to some performance regression, no? Could you try to run a micro-benchmark to check whether this performance impact is noticeable or not?

**Comment by llllllllll** [comment]
Thank you for taking the time to document this feature. While I love this code, it's definitely tricky.

Also, I think the original post meant to link to https://bugs.python.org/issue30486 instead of the nonlocal pep twice.

File: cloudpickle/cloudpickle.py (hunk 3 above)
Comment (pierreglaser) [reply cloudpickle/cloudpickle.py]: Yes indeed. On pickling this function
```python
def f():
    a = 1
    def g():
        return a
    return g

g = f()
```
this PR is 2.5% slower than master.

Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: Thanks.

File: cloudpickle/cloudpickle.py [hunk 4]
```diff
@@ -269,41 +269,81 @@ def func():
...
+    test and checker libraries decide to parse the whole file.
+    """
 
-           return _stub
+    if sys.version_info[:2] >= (3, 7):  # pragma: no branch
+        cell.cell_contents = value
+    else:
+        _cell_set = types.FunctionType(
+            _cell_set_template_code, {}, '_cell_set',
+            (), (cell,),)
```
Comment (ogrisel) [inline cloudpickle/cloudpickle.py]: style: this should fit on one line, no?

Comment (ogrisel) [reply cloudpickle/cloudpickle.py]:
```python
        _cell_set = types.FunctionType(
            _cell_set_template_code, {}, '_cell_set', (), (cell,),)
```

**Comment by pierreglaser** [comment]
@llllllllll 
> I think the original post meant to link to https://bugs.python.org/issue30486 instead of the nonlocal pep twice

Indeed, thanks!

**Comment by ogrisel** [comment]
@llllllllll anymore comment? Shall we merge?

**Comment by llllllllll** [comment]
LGTM

**Comment by ogrisel** [comment]
Merged! Thanks all.

**Pull request merged**